### PR TITLE
Various bug fixes

### PR DESCRIPTION
**Pull request by nwu63**
## Purpose
Fixed many bugs:
- The logic in `getModifiedOptions` was incorrect, this is now fixed.
- CaseInsensitive classes had bugs in implementation in a few places. In particular, `a.update(b)` did not work as expected when either `a` or `b` were not case insensitive. This is now fixed along with a couple of other things like comparisons.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
I added many new tests, which are now refactored to be more self-contained unit tests.

## Checklist
_Put an `x` in the boxes that apply._

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

### CODE REVIEW

**Review by sseraj — DISMISSED** [review]

**Review by marcomangano — APPROVED** [review]

**Review by eirikurj — APPROVED** [review]

**Comment by nwu63** [comment]
I converted this PR to draft because I realized there were a bunch of not-very-rigorous implementations, and in fact several untested features were not implemented correctly. In particular, comparisons and updates were not behaving as expected, those are now fixed.

I have also refactored the tests to be more like individual unit tests. There are now 32 tests! Hopefully this is more readable.

File: baseclasses/utils.py [hunk 1]
```diff
@@ -91,6 +103,11 @@ def __new__(self, *args, **kwargs):
         self.map = {}
         return super().__new__(self)
 
+    def __eq__(self, other):
```
Comment (marcomangano) [inline baseclasses/utils.py:106]: I am not really sure about the syntax and the role of this method

**Review by marcomangano — DISMISSED** [review]
Looks good, but I was not able to go over all the new tests. Maybe we can have a quick chat offline?

File: baseclasses/utils.py (hunk 1 above)
Comment (nwu63) [reply baseclasses/utils.py:106]: This function is used for comparisons, and is called when you run `obj1 == obj2`.

File: tests/test_CaseInsensitve.py [hunk 2]
```diff
@@ -7,61 +7,168 @@
...
-        value1 = 123
-        value2 = 321
-        value3 = 132
-        d = CaseInsensitiveDict({"OPtion1": value1})
+
+class TestCaseInsensitiveDict(unittest.TestCase):
+    def setUp(self):
+        self.d = CaseInsensitiveDict({"OPtion1": value1})
+        self.d2 = CaseInsensitiveDict({"opTION1": value2, "optioN2": value2})
+        self.d3 = CaseInsensitiveDict({"OPTION3": value3})
```
Comment (eirikurj) [inline tests/test_CaseInsensitve.py:19]: This seems to be unused

File: tests/test_CaseInsensitve.py [hunk 3]
```diff
@@ -7,61 +7,168 @@
...
+        self.assertIn("REGULAR DICT", self.d)
+        # check update preserves old capitalization
+        self.assertEqual(set(self.d.keys()), {"OPtion1", "regular dict"})
+
+    def test_update_regular_dict_with_dict(self):
+        # update regular dict with this dict
+        self.d4.update(self.d)
+        self.assertFalse(isinstance(self.d4, CaseInsensitiveDict))
+        self.assertTrue(isinstance(self.d4, dict))
+        self.assertNotIn("REGULAR DICT", self.d4)  # d3 is a dict and is not case insensitive
```
Comment (eirikurj) [inline tests/test_CaseInsensitve.py:89]: `d3` does not seem to be used here

**Review by eirikurj — CHANGES_REQUESTED** [review]
Looks good, very minor comments.

**Review by marcomangano — APPROVED** [review]

**Review by eirikurj — APPROVED** [review]

**Pull request merged**